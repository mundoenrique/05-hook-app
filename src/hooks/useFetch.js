import { useEffect, useRef, useState } from 'react';

export default function useFetch(url) {
	const isMounted = useRef(true);
	const [state, setState] = useState({});

	useEffect(() => {
		return () => {
			isMounted.current = false;
		};
	}, []);

	useEffect(() => {
		setState({
			loading: true,
			error: null,
			data: null,
		});
		fetch(url)
			.then((resp) => resp.json())
			.then((data) => {
				if (isMounted) {
					setState({
						loading: false,
						error: null,
						data,
					});
				}
			})
			.catch(() => {
				setState({
					loading: false,
					error: 'No fue posible obtener informaci√≥n',
					data: null,
				});
			});
	}, [url]);

	return state;
}
