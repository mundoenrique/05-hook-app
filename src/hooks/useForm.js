import { useState } from 'react';

export default function useForm(initialState = {}) {
	const [values, setValues] = useState(initialState);

	const resetForm = () => {
		setValues(initialState);
	};

	const handleInputChange = ({ target }) => {
		setValues({
			...values,
			[target.name]: target.value,
		});
	};

	return [values, resetForm, handleInputChange];
}
