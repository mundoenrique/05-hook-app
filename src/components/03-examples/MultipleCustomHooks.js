import { useCounter } from '../../hooks/useCounter';
import useFetch from '../../hooks/useFetch';

export default function MultipleCustomHooks() {
	const { counter, increment } = useCounter(1);
	const { loading, data } = useFetch(
		`https://breakingbadapi.com/api/quotes/${counter}`
	);
	const { author, quote } = !!data && data[0];

	return (
		<>
			<h3>Custom hooks</h3>
			<h4>Custom hooks</h4>
			<br />
			{loading ? (
				<div className="alert alert-info text-center">loading...</div>
			) : (
				<blockquote className="blockquote text-end">
					<p className="mb-2">{quote}</p>
					<footer className="blockquote-footer">{author}</footer>
				</blockquote>
			)}
			<button className="btn btn-outline-primary" onClick={() => increment()}>
				Siguiente frase
			</button>
		</>
	);
}
