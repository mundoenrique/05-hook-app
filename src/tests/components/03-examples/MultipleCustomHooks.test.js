import { shallow } from 'enzyme';
import MultipleCustomHooks from '../../../components/03-examples/MultipleCustomHooks';
import { useCounter } from '../../../hooks/useCounter';
import useFetch from '../../../hooks/useFetch';

jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');

describe('Pruebas sobre <MultipleCustomHooks />', () => {
	useCounter.mockReturnValue({
		counter: 10,
		increment: () => {},
	});

	test('debe renderizar correctamente <MultipleCustomHooks />', () => {
		useCounter.mockReturnValue({
			counter: 10,
			increment: () => {},
		});

		useFetch.mockReturnValue({
			loading: true,
			error: null,
			data: null,
		});

		const wrapper = shallow(<MultipleCustomHooks />);

		expect(wrapper).toMatchSnapshot();
	});

	test('Debe mostrar la informaciÃ³n', () => {
		const author = 'Enrique';
		const quote = 'Holamundo';

		useCounter.mockReturnValue({
			counter: 10,
			increment: () => {},
		});

		useFetch.mockReturnValue({
			loading: false,
			error: null,
			data: [
				{
					author: author,
					quote: quote,
				},
			],
		});

		const wrapper = shallow(<MultipleCustomHooks />);

		expect(wrapper.find('.alert').exists()).toBe(false);
		expect(wrapper.find('.mb-2').text().trim()).toBe(quote);
		expect(wrapper.find('.blockquote-footer').text().trim()).toBe(author);
	});
});
