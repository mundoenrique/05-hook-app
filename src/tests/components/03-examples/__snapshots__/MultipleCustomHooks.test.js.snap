// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre <MultipleCustomHooks /> debe renderizar correctamente <MultipleCustomHooks /> 1`] = `
<Fragment>
  <h3>
    Custom hooks
  </h3>
  <br />
  <div
    className="alert alert-info text-center"
  >
    loading...
  </div>
  <button
    className="btn btn-outline-primary"
    onClick={[Function]}
  >
    Siguiente frase
  </button>
</Fragment>
`;
